### PR TITLE
[mdatagen] Make metric builder track data points capacity by itself

### PR DESCRIPTION
**Pull request by dmitryax**
Scraper author should not care about the data points capacity. Metric builder can get the capacity by tracking number of data points added to the metrics. It makes the very first scrape a bit slower, but significantly simplifies the metric builder interface for scraper author.

Tracking issue: https://github.com/open-telemetry/opentelemetry-collector/issues/10904

### CODE REVIEW

File: cmd/mdatagen/metrics_v2.tmpl [hunk 1]
```diff
@@ -44,17 +44,33 @@ func DefaultMetricsSettings() MetricsSettings {
 	}
 }
 
+// metric holds data for generated metric and keeps track of data points slice capacity.
+type metric struct {
+	data     pdata.Metric // data buffer for generated metric.
+	capacity int          // max observed number of data points added to the metric.
+}
```
Comment (dmitryax) [inline cmd/mdatagen/metrics_v2.tmpl:51]: I'm thinking of making this class metric specific and moving record and emit function to this class. I'll prepare another refactoring PR once this one is merged

Comment (djaglowski) [reply cmd/mdatagen/metrics_v2.tmpl:51]: Emit func should probably be at the top level so it only needs to be called once. Unless we want to emit metrics one at a time, but this is not my expectation.

Comment (dmitryax) [reply cmd/mdatagen/metrics_v2.tmpl:51]: Exported Emit method will stay the same, here I mean just refactoring of internal methods

Comment (dmitryax) [reply cmd/mdatagen/metrics_v2.tmpl:51]: And actually I meant init not emit. updated commend

Comment (djaglowski) [reply cmd/mdatagen/metrics_v2.tmpl:51]: Ah, that makes sense. Thanks for clarifying.

**Review by djaglowski — APPROVED** [review]

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**